jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java

```diff
@@ -2306,9 +2316,15 @@ public String sniSelect(String keyType, Principal[] issuers, SSLSession session,
                     .filter(x509 -> x509.matches(sniHost))
                     .collect(Collectors.toList());
 
-                // No match, let the JDK decide unless unmatched SNIs are rejected.
                 if (matching.isEmpty())
-                    return isSniRequired() ? null : SniX509ExtendedKeyManager.SniSelector.DELEGATE;
+                {
+                    // There is no match for this SNI among the certificates valid for
+                    // this keyType; check if there is any certificate that matches this
+                    // SNI, as we will likely be called again with a different keyType.
+                    boolean anyMatching = aliasCerts().values().stream()
+                        .anyMatch(x509 -> x509.matches(sniHost));
+                    return isSniRequired() || anyMatching ? null : SniX509ExtendedKeyManager.SniSelector.DELEGATE;
```

Comment:
The problem with this algorithm is that if we have 5 matching certificates, this logic will be true just as much for the 5th as for calls 1 to 4.
I guess there is no where to remember a count easily?

Reply:
Good point. 
We can probably store an attribute in the handshake session -- not sure what though.
Initialize with all x509s, and remove those that don't match the `keyType`?

Reply:
sounds plausible. Something like:
 + Look for Integer attribute in session:
   * if not found, then count matching SNIs and set the Integer.
 + Decrement the integer
    * If == 0 return DELEGATE
    * else return null

Reply:
@gregw on further analysis, I think the first version is enough.

Assuming the server has [ec.domain.com keyType=EC, rsa.domain.com keyType=RSA], then:

- case 1: 
client wants to connect only with keyType=EC to rsa.domain.com;
we will be called with [ec.domain.com keyType=EC], which won't match; we know there is a matching certificate, we hope we will be called again (but we won't), so we return `null` (wrong), which fails the connection, but there is nothing we can do.

- case 2:
client wants to connect to unknown.domain.com;
we will be called with [ec.domain.com keyType=EC], which won't match; we figure out there is no match at all; we return either `null` or `delegate` (correct).

- default case:
there is a match for both keyType and SNI, so eventually we find the right alias to return.

The proposed fix improves case 1: before we were immediately returning `delegate`, now we return `null` in the hope to be called again. 
It's not perfect maybe, but if the client wants to connect with keyType=EC to a domain that does not support that keyType, I think failing the TLS handshake may be better than returning an EC certificate for a different domain.